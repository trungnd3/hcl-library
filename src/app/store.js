import { configureStore } from '@reduxjs/toolkit';
import authReducer from './slices/auth';
import bookReducer from './slices/book';

export const store = configureStore({
  reducer: {
    auth: authReducer,
    book: bookReducer,
  },
});
