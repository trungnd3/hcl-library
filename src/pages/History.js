import { Fragment, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import HistoryTable from '../components/Book/HistoryTable';
import HistoryFilter from '../components/Book/HistoryFilter';
import Centered from '../components/UI/Centered';

import { selectBorrowed } from '../app/action-creators/book';

const History = () => {
  const borrowedBooks = useSelector(selectBorrowed);
  const [filteredBooks, setFilteredBooks] = useState([]);

  useEffect(() => {
    setFilteredBooks(borrowedBooks);
  }, [borrowedBooks]);

  const filterHandler = (month, year) => {
    console.log(month, year);
    let filteredBorrowedBooks = [];
    if (!month.length && !year.length) {
      filteredBorrowedBooks = borrowedBooks;
    } else if (!!month && !year.length) {
      filteredBorrowedBooks = borrowedBooks.filter(
        (book) =>
          month === (new Date(book.borrowDate).getMonth() + 1).toString()
      );
    } else if (!!year && !month.length) {
      filteredBorrowedBooks = borrowedBooks.filter((book) => {
        return year === new Date(book.borrowDate).getFullYear().toString();
      });
    } else {
      filteredBorrowedBooks = borrowedBooks.filter((book) => {
        return (
          month === (new Date(book.borrowDate).getMonth() + 1).toString() &&
          year === new Date(book.borrowDate).getFullYear().toString()
        );
      });
    }

    setFilteredBooks(filteredBorrowedBooks);
  };

  return (
    <Fragment>
      <HistoryFilter onFilter={filterHandler} />
      {filteredBooks.length === 0 && (
        <Centered>
          <h3>No book found!</h3>
        </Centered>
      )}
      {filteredBooks.length > 0 && <HistoryTable books={filteredBooks} />}
    </Fragment>
  );
};

export default History;
