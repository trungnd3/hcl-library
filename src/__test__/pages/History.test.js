import React from 'react';
import History from '../../pages/History';
import { Provider } from 'react-redux';
import { store } from '../../app/store';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

let user;

describe('History component using RTL', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <History />
      </Provider>
    );
    user = userEvent.setup();
  });
  test('render snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <History />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should contain month label select', () => {
    const monthLabel = screen.getByText('Month');
    expect(monthLabel).toBeInTheDocument();
  });

  test('should correctly set default month', () => {
    expect(
      screen.getByRole('option', { name: 'Select a month' }).selected
    ).toBe(true);
  });

  test('should display the correct number of month options', () => {
    const monthSelect = screen.getByRole('month-filter');
    expect(monthSelect.children.length).toBe(13);
  });

  test('contains year label select', () => {
    const yearLabel = screen.getByText('Year');
    expect(yearLabel).toBeInTheDocument();
  });

  test('should correctly set default year', () => {
    expect(screen.getByRole('option', { name: 'Select a year' }).selected).toBe(
      true
    );
  });

  test('should display the correct number of year options', () => {
    const yearSelect = screen.getByRole('year-filter');
    expect(yearSelect.children.length).toBe(4);
  });

  test('contains Filter button', () => {
    const filterButton = screen.getByRole('button', { name: 'Filter' });
    expect(filterButton).toBeInTheDocument();
  });

  test('should display empty data', () => {
    const emptyText = screen.getByText('No book found!');
    expect(emptyText).toBeInTheDocument();
  });

  // test('submit form when click Filter button', async () => {
  //   const filterButton = screen.getByRole('button', { name: 'Filter' });
  //   const usernameInput = screen.getByRole('textbox', {
  //     name: 'Username',
  //   });
  //   const passwordInput = screen.getByRole('textbox', {
  //     name: 'Password',
  //   });

  //   fireEvent.change(usernameInput, { target: { value: 'abc' } });
  //   fireEvent.change(passwordInput, { target: { value: '123456' } });

  //   await user.click(submitButton);

  //   expect(localStorage.getItem('user')).toBe(usernameInput.value);
  // });

  // test('render History', () => {
  //   render(<History />);
  //   const historyText = screen.getByText('Welcome member!');
  //   expect(historyText).toBeInTheDocument();
  // });
});
