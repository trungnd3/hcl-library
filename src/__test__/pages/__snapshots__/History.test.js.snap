// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History component using RTL render snapshot 1`] = `
Array [
  <form
    className="m-3"
    onSubmit={[Function]}
  >
    <div
      className="row"
    >
      <div
        className="col-2"
      >
        <div>
          <div
            className="row"
          >
            <div
              className="col-3"
            >
              <label
                className="align-self-center me-2 w-50 text-end form-label"
                htmlFor="formMonth"
              >
                Month
              </label>
            </div>
            <div
              className="col"
            >
              <select
                className="form-select"
                id="formMonth"
                name="month"
                role="month-filter"
              >
                <option
                  value=""
                >
                  Select a month
                </option>
                <option
                  value={1}
                >
                  1
                </option>
                <option
                  value={2}
                >
                  2
                </option>
                <option
                  value={3}
                >
                  3
                </option>
                <option
                  value={4}
                >
                  4
                </option>
                <option
                  value={5}
                >
                  5
                </option>
                <option
                  value={6}
                >
                  6
                </option>
                <option
                  value={7}
                >
                  7
                </option>
                <option
                  value={8}
                >
                  8
                </option>
                <option
                  value={9}
                >
                  9
                </option>
                <option
                  value={10}
                >
                  10
                </option>
                <option
                  value={11}
                >
                  11
                </option>
                <option
                  value={12}
                >
                  12
                </option>
              </select>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col-2"
      >
        <div>
          <div
            className="row"
          >
            <div
              className="col-3"
            >
              <label
                className="align-self-center me-2 w-50 text-end form-label"
                htmlFor="formYear"
              >
                Year
              </label>
            </div>
            <div
              className="col"
            >
              <select
                className="form-select"
                id="formYear"
                name="year"
                role="year-filter"
              >
                <option
                  value=""
                >
                  Select a year
                </option>
                <option
                  value={2022}
                >
                  2022
                </option>
                <option
                  value={2021}
                >
                  2021
                </option>
                <option
                  value={2020}
                >
                  2020
                </option>
              </select>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col"
      >
        <div>
          <button
            className="btn btn-primary-library"
            disabled={false}
            type="submit"
          >
            Filter
          </button>
        </div>
      </div>
    </div>
  </form>,
  <div
    className="mt-5 text-center"
  >
    <h3>
      No book found!
    </h3>
  </div>,
]
`;
